boa_engine/src/object/jsmap.rs

```diff
@@ -0,0 +1,178 @@
+    builtins::Map,
+    object::{JsFunction, JsMapIterator, JsObject, JsObjectType, ObjectData},
+    Context, JsResult, JsValue,
+};
+
+use boa_gc::{Finalize, Trace};
+use std::ops::Deref;
+
+#[derive(Debug, Clone, Trace, Finalize)]
+pub struct JsMap {
+    inner: JsObject,
+}
+
+impl JsMap {
+    /// Create new Empty Map Object.
+    #[inline]
+    pub fn new(context: &mut Context) -> Self {
+        let map = Self::create_map(context);
+        Self { inner: map }
+    }
+
+    /// Create a new map object for any object that has a `@@Iterator` field.
+    #[inline]
+    pub fn from_js_iterable(iterable: &JsValue, context: &mut Context) -> JsResult<Self> {
+        // Create a new map object.
+        let map = Self::create_map(context);
+
+        // Let adder be Get(map, "set") per spec. This action should not fail with default map.
+        let adder = map
+            .get("set", context)
+            .expect("creating a map with the default prototype must not fail");
+
+        let _completion_record = add_entries_from_iterable(&map, iterable, &adder, context)?;
+
+        Ok(Self { inner: map })
+    }
+
+    /// Create a [`JsMap`] from a [`JsObject`], if the object is not a `Map`, throw a `TypeError`.
+    #[inline]
+    pub fn from_object(object: JsObject, context: &mut Context) -> JsResult<Self> {
+        if object.borrow().is_map() {
+            Ok(Self { inner: object })
+        } else {
+            context.throw_type_error("object is not a Map")
+        }
+    }
+
+    // Utility function to generate default Map object.
+    fn create_map(context: &mut Context) -> JsObject {
+        // Get default Map prototype
+        let prototype = context.intrinsics().constructors().map().prototype();
+
+        // Create a default map object with [[MapData]] as a new empty list
+        JsObject::from_proto_and_data(prototype, ObjectData::map(OrderedMap::new()))
+    }
+
+    /// Return a new Iterator object that contains the [key, value] pairs in order of assertion.
+    #[inline]
+    pub fn entries(&self, context: &mut Context) -> JsResult<JsMapIterator> {
+        let iterator_record = Map::entries(&self.inner.clone().into(), &[], context)?
+            .get_iterator(context, None, None)?;
+        let map_iterator_object = iterator_record.iterator();
+        JsMapIterator::from_object(map_iterator_object.clone(), context)
+    }
+
+    /// Return the keys Iterator object
+    #[inline]
+    pub fn keys(&self, context: &mut Context) -> JsResult<JsMapIterator> {
+        let iterator_record = Map::keys(&self.inner.clone().into(), &[], context)?
+            .get_iterator(context, None, None)?;
+        let map_iterator_object = iterator_record.iterator();
+        JsMapIterator::from_object(map_iterator_object.clone(), context)
+    }
+
+    /// Insert a new entry into the Map object
+    #[inline]
+    pub fn set<T>(&self, key: T, value: T, context: &mut Context) -> JsResult<JsValue>
+    where
+        T: Into<JsValue>,
```

Comment:
This has the limitation of both the `key` and the `value` needing to be of the same type. We don't really need this limitation. I would say, we should change the definition of the function:
```suggestion
    pub fn set<K, V>(&self, key: K, value: V, context: &mut Context) -> JsResult<JsValue>
    where
        K: Into<JsValue>,
        V: Into<JsValue>,
```

Reply:
Nice catch! I updated the types for `key` and `value`